### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by frankhillard**
Transpilation of the Permit-cameligo smart contract in JsLigo

### CODE REVIEW

File: src/extension.jsligo [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    counter: nat,
+    defaultExpiry: seconds,
+    maxExpiry: seconds,
+    permits: permits,
+    userExpiries: userExpiries,
+    permitExpiries: permitExpiries,
+    tokenTotalSupply : tokenTotalSupply
+};
+
+let get_user_defined_expiry = ([from_, ext]: [address, t]) : option<seconds> => {
```
Comment (d-plaindoux) [inline src/extension.jsligo]: `const` should be preferred instead of `let` in JsLIGO (Warning raised during the compilation)

File: README.md [hunk 2]
```diff
@@ -1 +1,151 @@
...
+
+For deploy scripts, you also need to have [nodejs](https://nodejs.org/en/) installed,
+up to version 14 and docker if you wish to deploy on a sandbox.
+
+## Usage
+
+1. Run `make install` to install dependencies
+2. Run `make` to see available commands
+
+## Use case: taco shop loyalty program
```
Comment (d-plaindoux) [inline README.md:33]: The explanation is given with the taco shop as a use case and the implementation uses a generic FA2. 
When I read the documentation I'm waiting for a taco shop extension.

File: src/extension.jsligo [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    return match (Big_map.find_opt(from_, ext.userExpiries), {
+        None: () => (Some(ext.defaultExpiry)),
+        Some: (exp: option<seconds>) => exp
+    });
+};
+
+let getExpiry = ([ext, paramPermitKey]: [t, permitKey]) : seconds => {
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Function syntax should be reviewed and replaced by a much more idiomatic syntax (for JsLIGO):
```
const getExpiry = (ext: t, paramPermitKey permitKey) : seconds => {
```

File: src/extension.jsligo [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+export let assertAdmin = (ext : t) : unit =>
+    assert_with_error((Tezos.get_sender() == ext.admin), Errors.requiresAdmin);
+
+export let setAdmin = ([ext, admin] : [t, address]) : t => {
+    assertAdmin(ext);
+    return { ...ext, admin : admin };
+};
+
+let getSupply = ([supply, tokenId] : [tokenTotalSupply, FA2.Ledger.token_id]) : nat =>
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Why have we such a strong link between the extension implementation and the FA2?

File: src/extension.jsligo [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    match (Big_map.find_opt(tokenId, supply), { 
+        Some: (a: nat) => a,
+        None: () => (failwith(FA2.Errors.undefined_token) as nat)
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Should not use FA2 messages

File: src/main.jsligo [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                    (Some(seconds)),
+                    s.extension.permitExpiries)
+                }}
+            ]
+        }
+    });
+};
+
+// TZIP-17 implementation of TZIP-12 Transfer
+let transferPermitted = ([t, s]: [FA2.transfer, storage]) : result => {
```
Comment (d-plaindoux) [inline src/main.jsligo]: If we break the dependency between `extensions` and FA2 the permission and the transfer should be composed providing a `transferPermitted` behaviour. This should also prevent code duplication.

**Review by d-plaindoux — CHANGES_REQUESTED** [review]
A clean separation between TZIP-17 and FA2 should be possible.

File: src/main.jsligo (hunk 6 above)
Comment (frankhillard) [reply src/main.jsligo]: agreed, `transferPermitted` dependency to FA2 has been removed

File: src/extension.jsligo (hunk 5 above)
Comment (frankhillard) [reply src/extension.jsligo]: agreed Supply managment has been moved to a separate file

File: src/extension.jsligo (hunk 4 above)
Comment (frankhillard) [reply src/extension.jsligo]: agreed Supply managment has been moved to a separate file

File: src/extension.jsligo (hunk 1 above)
Comment (frankhillard) [reply src/extension.jsligo]: agreed , done

File: src/extension.jsligo (hunk 3 above)
Comment (frankhillard) [reply src/extension.jsligo]: disagree, The documentation specifies we should use `const getExpiry =[ext, paramPermitKey] : [t, permitKey] : seconds => `

File: src/extension.jsligo [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    userExpiries: userExpiries,
+    permitExpiries: permitExpiries,
+    extension : a
+};
+
+const get_user_defined_expiry : <T>((from_: address, ext: t<T>) => option<seconds>) = 
+(from_: address, ext: t<T>) => {
+    return match (Big_map.find_opt(from_, ext.userExpiries), {
+        None: () => (Some(ext.defaultExpiry)),
+        Some: (exp: option<seconds>) => exp
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Explicit type can be removed (inferred)

```suggestion
        Some: exp => exp
```

File: src/extension.jsligo [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    });
+};
+
+const getExpiry : <T>((ext: t<T>, paramPermitKey: permitKey) => seconds) = 
+(ext: t<T>, paramPermitKey: permitKey) => {
+    match ((Big_map.find_opt([paramPermitKey[0], paramPermitKey[1]], ext.permitExpiries) as option<option<seconds>>), {
+        None: () => {
+                match ((Big_map.find_opt(paramPermitKey[0], ext.userExpiries) as option<option<seconds>>), {
+                    None: () => ext.defaultExpiry,
+                    Some: (exp: option<seconds>) => {
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Explicit type can be removed (inferred)
```suggestion
                    Some: exp => {
```

File: src/extension.jsligo [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                        match (exp, {
+                            None: () => ext.defaultExpiry,
+                            Some: (t: seconds) => t
```
Comment (d-plaindoux) [inline src/extension.jsligo]:
```suggestion
                            Some: t => t
```

File: src/extension.jsligo [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                        })
+                    }
+                })
+            },
+        Some: (p: option<seconds>) => {
+                match (p, {
+                    None: () => ext.defaultExpiry,
+                    Some: (exp: seconds) => exp
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Explicit type can be removed (inferred)
```suggestion
                    Some: exp => exp
```

File: src/extension.jsligo [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        });
+    if (permitSubmitTime == (0 as timestamp)) {
+        return [false, ext]
+    } else {
+        let effectiveExpiry =
+            match (Big_map.find_opt(paramPermitKey, ext.permitExpiries), {
+                None: () => {
+                    match (Big_map.find_opt(paramPermitKey[0], ext.userExpiries), {
+                        None: () => (Some(ext.defaultExpiry)),
+                        Some: (exp: option<seconds>) => exp
```
Comment (d-plaindoux) [inline src/extension.jsligo]:
```suggestion
                        Some: exp => exp
```

File: src/extension.jsligo [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+                Some: (exp: option<seconds>) => exp
+            });
+        return match (effectiveExpiry, {
+            None: () => (failwith(Errors.missingExpiry) as [bool, t<T>]),
+            Some: (effective_exp: seconds) => {  
+                let permits = Big_map.remove(paramPermitKey, ext.permits);
+                let isAuthorised = abs ((Tezos.get_now()) - permitSubmitTime) < effective_exp;
+                [isAuthorised, { ...ext, permits : permits }]
+            }
+        });
```
Comment (d-plaindoux) [inline src/extension.jsligo]: Explicit types can be removed (inferred)

```suggestion
        return match (effectiveExpiry, {
            None: () => failwith(Errors.missingExpiry),
            Some: effective_exp => {  
                let permits = Big_map.remove(paramPermitKey, ext.permits);
                let isAuthorised = abs ((Tezos.get_now()) - permitSubmitTime) < effective_exp;
                [isAuthorised, { ...ext, permits : permits }]
            }
        });
```

File: src/main.jsligo [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        Transfer:         (p: FA2.transfer) => transferPermitted(p,s),
+        Balance_of:       (p: FA2.balance_of) => FA2.balance_of(p)(s),
+        Update_operators: (p: FA2.update_operators) => FA2.update_ops(p)(s),
+        Create_token:     (p: [create_token, address, nat]) => create(p, s),
+        Mint_token:       (p: list<mintOrBurn>) => mint(p, s),
+        Burn_token:       (p: list<mintOrBurn>) => burn(p, s),
+        Permit:           (p: list<permitParams>) => permit(p, s),
+        SetExpiry:        (p: expiryParams) => setExpiry(p, s),
+        SetAdmin:         (p: address) => setAdmin(p, s)
+    })
```
Comment (d-plaindoux) [inline src/main.jsligo]: Explicit types can be removed.

Comment (melwyn95) [reply src/main.jsligo]: Will do

Comment (melwyn95) [reply src/main.jsligo]: Done

File: README.md (hunk 2 above)
Comment (melwyn95) [reply README.md:33]: Closing this, as I want to merge this PR

**Review by melwyn95 — APPROVED** [review]

**Pull request merged**